Mathlib/Topology/Instances/AddCircle/Defs.lean

```diff
@@ -534,14 +528,29 @@ theorem card_addOrderOf_eq_totient {n : ℕ} :
       n.totient_eq_card_lt_and_coprime]
     simp only [Nat.gcd_comm]
 
+omit [Fact (0 < p)]
+
+theorem finite_torsion {n : ℕ} (hn : 0 < n) : { u : AddCircle p | n • u = 0 }.Finite := by
+  wlog hp : 0 < p generalizing p
+  on_goal 2 =>
+    have := Fact.mk hp
+    convert Set.finite_range (fun m : Fin n ↦ (↑(↑m / ↑n * p) : AddCircle p))
+    simp_rw [nsmul_eq_zero_iff hn, range, Fin.exists_iff, exists_prop]
+  obtain rfl | hp := eq_or_lt_of_not_gt hp
+  · refine (Set.finite_singleton 0).subset fun u hu ↦ ?_
+    obtain ⟨u, rfl⟩ := QuotientAddGroup.mk_surjective u
+    rw [Set.mem_setOf, ← coe_nsmul] at hu
+    rw [Set.mem_singleton_iff]
+    rw [QuotientAddGroup.eq_zero_iff, zmultiples_zero_eq_bot] at hu ⊢
+    exact (_root_.nsmul_eq_zero_iff hn.ne').mp hu
+  dsimp only [AddCircle] at this ⊢
+  convert ← this (-p) (neg_pos.mpr hp) using 1
+  · rw [zmultiples_neg]
+  exact congr_arg_heq (fun H ↦ {u : 𝕜 ⧸ H | n • u = 0}) zmultiples_neg
```

Comment:
This uses an equality of types, which always makes me nervous. Can you say why you chose this proof/why there is no better way?